### ISSUE TITLE
How to ensemble models?

### ISSUE THREAD

**Issue by ylqfp** [fchollet/keras#2775]
Hi,
I'm trying to ensemble different model to achieve better performace, how could i combine them together?
The only way I could think of is a naive method:
``` python
model1 = sequential()
model1.compile()
model1.fit()
model2 = sequential()
model2.compile()
model2.fit()
model3 = sequential()
model3.compile()
model3.fit()

prob1 = model1.predict_prob()
prob2 = model2.predict_prob()
prob3 = model3.predict_prob()
final_prob = (prob1 + prob2 + prob3)/3

```
Is this OK in practical?
Is there any end-to-end ensemble method?
Thanks!

**Comment by jskDr** [comment]
It is exactly what I am doing in my task. I just have question on this. Do
you use the same training set and the same validation set for all models?

On Sun, May 22, 2016 at 5:48 AM Leo YU <notifications@github.com> wrote:

> Hi,
> I'm trying to ensemble different model to achieve better performace, how
> could i combine them together?
> The only way I could think of is a naive method:
>
> model1 = sequential()
> model1.compile()
> model1.fit()
> model2 = sequential()
> model2.compile()
> model2.fit()
> model3 = sequential()
> model3.compile()
> model3.fit()
>
> prob1 = model1.predict_prob()
> prob2 = model2.predict_prob()
> prob3 = model3.predict_prob()
> final_prob = (prob1 + prob2 + prob3)/3
>
> Is this OK in practical?
> Is there any end-to-end ensemble method?
> Thanks!
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/fchollet/keras/issues/2775>
>
-- 

Sungjin (James) Kim, PhD

Postdoc, CCB in Harvard

sungjinkim@fas.harvard.edu

[Web] http://aspuru.chem.harvard.edu/james-sungjin-kim/

[Linkedin] https://www.linkedin.com/in/jamessungjinkim

[Facebook] https://www.facebook.com/jamessungjin.kim

[alternative email] jamessungjin.kim@gmail.com

**Comment by lemuriandezapada** [comment]
you can add a merge layer on those outputs with mode="ave"
Doing it with the functional API would be much better

**Comment by ylqfp** [comment]
Great! Thanks!

**Comment by alyato** [comment]
@lemuriandezapada  Does the functional API  do it well? I see some guys built their models using it .
why not use the Graph model?

**Comment by alyato** [comment]
@ylqfp Do you share a simple example how  to implement by the functional API? Thanks.

**Comment by guotong1988** [comment]
same question

**Comment by soon-will** [comment]
I have the same question, could you show how do you solve it in the end.

**Comment by colt18** [comment]
You should always use the same training and validation sets for fitting your models. Else it'll yield to a data  leak. 
@jskDr

**Comment by leoliveira00** [comment]
This is a good tutorial on the subject:
https://towardsdatascience.com/ensembling-convnets-using-keras-237d429157eb

**Comment by HarleysZhang** [comment]
> This is a good tutorial on the subject:
> https://towardsdatascience.com/ensembling-convnets-using-keras-237d429157eb

thanks for you share!

**Comment by Deepanp8** [comment]
> you can add a merge layer on those outputs with mode="ave"
> Doing it with the functional API would be much better

any ensemble classifier is available Other than average ensembler? if available means give suggestion...
